Implementation of Parallelization to analysis.gnm

talagayev opened the pull request:
Fixes #4672 

Changes made in this Pull Request:

- Parallelization of the class `GNMAnalysis` in `analysis.gnm.py`
- Addition of parallelization tests (including fixtures in analysis/conftest.py)
- update of `CHANGELOG`


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4700.org.readthedocs.build/en/4700/

<!-- readthedocs-preview mdanalysis end -->

talagayev committed:
Update gnm.py

talagayev committed:
Update conftest.py

Added GNM

talagayev committed:
Update test_gnm.py

Added GNM client into tests

talagayev committed:
Update CHANGELOG

Added Change to Changelog

talagayev committed:
Update CHANGELOG

added name to changelog

pep8speaks commented:
Hello @talagayev! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/analysis/gnm.py`](https://github.com/MDAnalysis/mdanalysis/blob/b84fc1ab8584c37c7d354b19fa8290e5ebc3f027/package/MDAnalysis/analysis/gnm.py):

> [Line 250:80](https://github.com/MDAnalysis/mdanalysis/blob/b84fc1ab8584c37c7d354b19fa8290e5ebc3f027/package/MDAnalysis/analysis/gnm.py#L250): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 251:78](https://github.com/MDAnalysis/mdanalysis/blob/b84fc1ab8584c37c7d354b19fa8290e5ebc3f027/package/MDAnalysis/analysis/gnm.py#L251): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 260:1](https://github.com/MDAnalysis/mdanalysis/blob/b84fc1ab8584c37c7d354b19fa8290e5ebc3f027/package/MDAnalysis/analysis/gnm.py#L260): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace







##### Comment last updated at 2024-09-09 17:42:45 UTC

talagayev commented:
overall the tests all work for the exception of `test_gnm_SVD_fail` due to the parallelization leading to an error for this line:
`with pytest.warns(UserWarning, match=msg):`

with the error being:
`E           Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.`
`E           Emitted warnings: []`

hmacdope commented:
@orbeckst would you be able to review here?

orbeckst committed:
Merge branch 'develop' into gnm_parallel

orbeckst commented on package/CHANGELOG:
```diff
@@ -53,6 +54,7 @@ Fixes
  * Fix groups.py doctests using sphinx directives (Issue #3925, PR #4374)
 
 Enhancements
+ * Enables parallelization for analysis.gnm.GNMAnalysis (Issue #4672)
```
We customarily put new entries first but for anything related to parallelization I propose to add them _below_ the entry for Issue #4158, just so that there is more context.

talagayev replied:
Yep, I can adjust the Changelog to add them below Issue #4158 :)

orbeckst reviewed (changes requested):
Azure has timeout errors in multiprocessing 
```
=================================== FAILURES ===================================
_ Test_Backends.test_all_backends_give_correct_results[square-iterable1-answer1] _
...
self = <multiprocessing.popen_fork.Popen object at 0x7efe2d29ba30>, flag = 0

    def poll(self, flag=os.WNOHANG):
        if self.returncode is None:
            try:
>               pid, sts = os.waitpid(self.pid, flag)
E               Failed: Timeout >200.0s
```
I am going to assume that this is related to #4209 so I don't think that this PR can do anything about it. 

All other tests pass for me. I didn't see a failure for `test_gnm_SVD_fail`. 

## Changes
For actual changes:
* add a versionchanged to the GNMAnalysis class to state that parallelization is now enabled
* change the entry in CHANGELOG

## Comment on test duration increase
Note: The GNM tests are some of the slowest ones that are running:
```
============================= slowest 50 durations =============================
9.90s call     MDAnalysisTests/analysis/test_gnm.py::test_gnm[client_GNMAnalysis0]
8.92s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-None-50-2-25]
8.81s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-20-50-2-15]
8.79s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-20-50-None-30]
8.67s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-None-None-2-49]
8.57s call     MDAnalysisTests/analysis/test_gnm.py::test_gnm[client_GNMAnalysis1]
8.24s call     MDAnalysisTests/coordinates/test_pdb.py::test_conect_bonds_all
8.08s call     MDAnalysisTests/analysis/test_contacts.py::TestContacts::test_n_initial_contacts[datafiles1-41814]
8.06s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis[client_GNMAnalysis1]
7.67s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis_weights_None[client_GNMAnalysis1]
7.36s call     MDAnalysisTests/analysis/test_encore.py::TestEncore::test_ces_error_estimation_ensemble_bootstrap
7.11s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis[client_GNMAnalysis0]
6.99s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis_weights_None[client_GNMAnalysis0]
6.93s call     MDAnalysisTests/utils/test_pickleio.py::test_fileio_pickle
6.14s call     MDAnalysisTests/coordinates/test_gro.py::TestGROLargeWriter::test_writer_large
5.70s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis[client_GNMAnalysis2]
5.64s call     MDAnalysisTests/analysis/test_gnm.py::test_gnm_run_step[client_GNMAnalysis1]
```

Previously (from PR #4682), GNM was slow but not the slowest.
```
============================= slowest 50 durations =============================
9.49s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-20-50-None-30]
9.02s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-20-50-2-15]
8.90s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-None-None-2-49]
8.63s call     MDAnalysisTests/coordinates/test_pdb.py::test_conect_bonds_all
8.58s call     MDAnalysisTests/analysis/test_base.py::test_AnalysisFromFunction[client_AnalysisFromFunction1-None-50-2-25]
8.11s call     MDAnalysisTests/analysis/test_contacts.py::TestContacts::test_n_initial_contacts[datafiles1-41814]
7.41s call     MDAnalysisTests/analysis/test_encore.py::TestEncore::test_ces_error_estimation_ensemble_bootstrap
6.97s call     MDAnalysisTests/utils/test_pickleio.py::test_fileio_pickle
6.26s call     MDAnalysisTests/coordinates/test_gro.py::TestGROLargeWriter::test_writer_large
5.79s call     MDAnalysisTests/analysis/test_rms.py::TestRMSD::test_weights_mass_is_mass_weighted[client_RMSD1]
5.73s call     MDAnalysisTests/analysis/test_rms.py::TestRMSD::test_rmsd_atomgroup_selections[client_RMSD1]
5.60s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis_weights_None
5.44s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis
```

It looks as if the tests in parallel take longer than the serial tests — possibly setup and then competition with the parallel runners.

talagayev commented:
> All other tests pass for me. I didn't see a failure for `test_gnm_SVD_fail`.


Yes for this PR I didn't add the `client_GNMAnalysis` to the `test_gnm_SVD_fail` pytest, since I was not sure if it is required to have it and without it all the remaining tests did work without errors, thus I was not sure if it should be implemented if it will cause an error 😿

talagayev commented:
As for the other classes, since I tried also adding the parallelization implementation to them, is there a certain way to say when they are unparallelizable? since I am quite hesitant to mark them as unparallelizable, since maybe I could overlook an option to implement the parallelization with some adjustments.

talagayev committed:
Update CHANGELOG

moved the parallelization below Issue #4158

talagayev committed:
Update gnm.py

Addiditon of versionchanged with the mention of the support for multiprocessing and dask backends

orbeckst commented on package/MDAnalysis/analysis/gnm.py:
```diff
@@ -245,8 +245,18 @@ class GNMAnalysis(AnalysisBase):
        store results as attributes ``times``, ``eigenvalues`` and
        ``eigenvectors`` of the ``results`` attribute.
+
+    .. versionchanged:: 2.8.0
+       introduced a :meth:`get_supported_backends` allowing for execution on with
+       ``multiprocessing`` and ``dask`` backends.
```
fix grammar and make it more obvious to users what happened (I know, this is a copy of the RMSD one... should have caught it then — perhaps fix it, too in this PR?)
```suggestion
    .. versionchanged:: 2.8.0
       Enabled **parallel execution** with the ``multiprocessing`` and ``dask`` 
       backends; use the new method :meth:`get_supported_backends` to see all 
       supported backends.
```

orbeckst reviewed (changes requested):
I am ok with not testing parallelization in `test_gnm_SVD_fail`.

Can you please update the versionchanged (see my suggestion)?

Otherwise looking ready to me.

talagayev committed:
Update package/MDAnalysis/analysis/gnm.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

talagayev commented:
> I am ok with not testing parallelization in `test_gnm_SVD_fail`.
> 
> Can you please update the versionchanged (see my suggestion)?
> 
> Otherwise looking ready to me.

Nice, then we can leave it like this for the tests :)

Yes adjusted it, yes was just copy-pasting from the RMSD class versionchanged :)

orbeckst commented:
> As for the other classes, since I tried also adding the parallelization implementation to them, is there a certain way to say when they are unparallelizable? since I am quite hesitant to mark them as unparallelizable, since maybe I could overlook an option to implement the parallelization with some adjustments.

If you're unsure, start a discussion on the issue. There are plenty of people around who can look at an algorithm and given an opinion. Perhaps sometimes we come up with an idea how to do it. For example, RMSF is in principle parallelizable with split-apply-combine, one just has to do some extra work (see https://www.mdanalysis.org/pmda/api/rmsf.html ) but at the moment only serial is enable https://github.com/MDAnalysis/mdanalysis/blob/7618e05176b5f2c3008f375f840396ea8592488a/package/MDAnalysis/analysis/rms.py#L756-L757 because someone has to do the extra work.

talagayev commented:
> > As for the other classes, since I tried also adding the parallelization implementation to them, is there a certain way to say when they are unparallelizable? since I am quite hesitant to mark them as unparallelizable, since maybe I could overlook an option to implement the parallelization with some adjustments.
> 
> If you're unsure, start a discussion on the issue. There are plenty of people around who can look at an algorithm and given an opinion. Perhaps sometimes we come up with an idea how to do it. For example, RMSF is in principle parallelizable with split-apply-combine, one just has to do some extra work (see https://www.mdanalysis.org/pmda/api/rmsf.html ) but at the moment only serial is enable
> 
> https://github.com/MDAnalysis/mdanalysis/blob/7618e05176b5f2c3008f375f840396ea8592488a/package/MDAnalysis/analysis/rms.py#L756-L757
> 
> because someone has to do the extra work.

Ah I see, that sounds good, since I was sure for the implementation for this class and in `analysis.bat` and from the testing of parallelization implementation `analysis.align` it looks that  `analysis.align` is unparallelizable due to getting the following error:
`ValueError: Can not parallelize class <class 'MDAnalysis.analysis.align.AlignTraj'>`

so that it has the following case which is written in the `base.py`:

`if not self.parallelizable and backend_class is not BackendSerial:`
`    raise ValueError(f"Can not parallelize class {self.__class__}")`

Thus I would tend to mark it as unparallelizable and create a PR for it :)

orbeckst reviewed (approved):
Looks ready to me.

The Azure failure is a multiprocessing timeout failure and likely not due to this PR.

A concern is the increased run time of the tests but that's a general thing with the parallel tests, I think.

orbeckst commented:
Thanks @talagayev !

talagayev commented:
> Thanks @talagayev !

Happy to help :)
